src/sage/rings/lazy_series_ring.py

```diff
@@ -659,6 +659,324 @@ def undefined(self, valuation=None):
+
+            sage: L.<z> = LazyPowerSeriesRing(QQ)
+            sage: A = L.undefined()
+            sage: B = L.undefined()
+            sage: FA = A^2 + B^2 - 2 - z*B
+            sage: FB = B^3 + 2*A^3 - 3 - z*(A + B)
+            sage: L.define_implicitly([(A, [1]), (B, [1])], [FA, FB])
+            sage: A^2 + B^2 - 2 - z*B
+            O(z^7)
+            sage: B^3 + 2*A^3 - 3 - z*(A + B)
+            O(z^7)
+
+            sage: L.<z> = LazyPowerSeriesRing(QQ)
+            sage: A = L.undefined(valuation=3)
+            sage: B = L.undefined(valuation=2)
+            sage: C = L.undefined(valuation=2)
+            sage: FA = (A^2 + B^2)*z^2
+            sage: FB = A*B*z
+            sage: FC = (A + B + C)*z^2
+            sage: L.define_implicitly([A, B, C], [FA, FB, FC])
+            sage: A
+            O(z^10)
+            sage: B
+            O(z^16)
+            sage: C
+            O(z^23)
+
+            sage: L.<z> = LazyPowerSeriesRing(QQ)
+            sage: A = L.undefined()
+            sage: B = L.undefined()
+            sage: C = L.undefined()
+            sage: L.define_implicitly([A, B, C], [B - C - 1, B*z + 2*C + 1, A + 2*C + 1])
+            sage: A + 2*C + 1
+            O(z^7)
+
+        The following system does not determine `B`, but the solver
+        will inductively discover that each coefficient of `A` must
+        be zero.  Therefore, asking for a coefficient of `B` will
+        loop forever::
+
+            sage: L.<z> = LazyPowerSeriesRing(QQ)
+            sage: A = L.undefined()
+            sage: B = L.undefined()
+            sage: C = L.undefined()
+            sage: D = L.undefined()
+            sage: L.define_implicitly([(A, [0,0,0]), (B, [0,0]), (C, [0,0]), (D, [0,0])], [C^2 + D^2, A + B + C + D, A*D])
+            sage: B[2]  # known bug, not tested
+
+        A bivariate example::
+
+            sage: R.<z,q> = LazyPowerSeriesRing(QQ)
+            sage: g = R.undefined()
+            sage: R.define_implicitly([g], [g - (z*q + z*g*~(1-g))])
+            sage: g
+            z*q + z^2*q + z^3*q + (z^4*q+z^3*q^2) + (z^5*q+3*z^4*q^2) + O(z,q)^7
+
+        A bivariate example involving composition of series::
+
+            sage: R.<z,q> = LazyPowerSeriesRing(QQ)
+            sage: g = R.undefined()
+            sage: R.define_implicitly([g], [g - (z*q + z*g*~(1-g))])
+            sage: g
+            z*q + z^2*q + z^3*q + (z^4*q+z^3*q^2) + (z^5*q+3*z^4*q^2) + O(z,q)^7
+
+        The following does not work currently, because the equations
+        determining the coefficients come in bad order::
+
+            sage: L.<x,y,t> = LazyPowerSeriesRing(QQ)
+            sage: M1 = L.undefined()
+            sage: M2 = L.undefined()
+            sage: eq1 = t*x*y*M2(0, 0, t) + (t - x*y)*M1(x, y, t) + x*y - t*M1(0, y, t)
+            sage: eq2 = (t*x-t)*M2(0, y, t) + (t - x*y)*M2(x, y, t)
+            sage: L.define_implicitly([M1, M2], [eq1, eq2])
+            sage: M1[1]  # known bug, not tested
```

Comment:
I am finally trying to debug the example above.  It currently actually doesn't fail as claimed, but rather because of a problem in the composition of power series.  The individual coefficients of a (multivariate) composition are computed in line 5272ff of `lazy_series.py`:
```
        def coefficient(n):
            r = R.zero()
            for i in range(n // gv + 1):
                c = coeff_stream[i]
                if c in self.base_ring():
                    c = P(c)
                    r += c[n]
                else:
                    r += c(g)[n]
            return r
```
In the case at hand, when asking for `M1[0]`, we eventually have `c` equal `FESDUMMY_1` in the `Multivariate Polynomial Ring in x, y, t over Fraction Field of Infinite polynomial ring in FESDUMMY over Rational Field`, and `g` is the triple `[x, y, t]` of `Multivariate Lazy Taylor Series Ring in x, y, t over Rational Field`.

One might hope that `c(g)` equals `FESDUMMY_1` in `Multivariate Lazy Taylor Series Ring in x, y, t over Fraction Field of Infinite polynomial ring in FESDUMMY over Rational Field`,  but it does not.

Instead it gives `y` in `Fraction Field of Infinite polynomial ring in FESDUMMY over Multivariate Lazy Taylor Series Ring in x, y, t over Rational Field`.

Reply:
The responsible code in `multi_polynomial_element.MPolynomial_element` effectively adds `0` from `Multivariate Lazy Taylor Series Ring in x, y, t over Rational Field` and `c`, and somewhat surprisingly obtains `y` in `Fraction Field of Infinite polynomial ring in FESDUMMY over Multivariate Lazy Taylor Series Ring in x, y, t over Rational Field`.

Reply:
I guess that this can be circumvented by being much more explicit in the local `coefficient` routine mentioned above, but I am worried about obscure bugs elsewhere - in particular in user code.

Reply:
See #37756 for the siliently wrong result.

I just noticed that there is currently no example using multivariate composition.  I guess they would all fail.